frontend/server/handlers/artifacts.ts

```diff
@@ -239,6 +260,62 @@ function getGCSArtifactHandler(options: { key: string; bucket: string }, peek: n
   };
 }
 
+function getVolumeArtifactsHandler(options: { bucket: string; key: string }, peek: number = 0) {
+  const { key, bucket } = options;
+  return async (req: Request, res: Response) => {
+    const filePath = path.join('/', key);
```

Comment:
discussion: shall we get filePath combining volume mount path and key (key is just relative path in the volume)

Reply:
I think I get your point, but how to handle below case:
```
spec:
  template:
    spec:
      containers:
        volumeMounts:
          - mountPath: /data1
            subPath: pipeline1
            name: output
          - mountPath: /data2
            subPath: pipeline2
            name: output
      volumes:
        - name: output
          persistentVolumeClaim:
                claimName: output,
                readOnly: true
```
just use `volume://output/for/bar/path` info can't find which `output` volume mount to use. but `volume://output/data1/for/bar/path` can. WDYT?

Reply:
Good point! I don't have experience administering a kubernetes cluster, is subPath a common requirement?

My original proposal was based on the reasoning:

Because the volumeName + key combination in mlpipeline-ui-metadata.json needs to be persisted as permanent metadata. I was hoping they can be as globally unique as possible. The more options we give users to configure, the more knowledge users need to configure them correctly.

The key can be opinionated to only accept relative path from volume root and we will only allow mounting a volume to `ml-pipeline-ui` in one mountPath without subPath. Do you think it would be enough for daily usage? What limitations do you foresee? If the flexibility is not enough, we should consider alternatives.


Regarding your proposal, I was worrying key must be an absolute path on pods that try to mount the volumes and read them. So it will be users' responsibility to choose a key that won't have conflicts with paths used in those pods. e.g. we may decide to use a new path in those pods in the future, all existing users' keys having conflict with them will be a problem.
To solve this problem, can we have a convention like
```
absolute path in ml-pipeline-ui pod = /etc/artifacts/<key>
```
or whatever prefix you think would be more appropriate.

Another benefit is, with your original proposal, users can specify a <key> to try to read arbitrary files from `ml-pipeline-ui` pod, they might use this mechanism to steal service account tokens etc. Therefore, we needed to check whether specified <key> is a subpath of a mounted volume. With a fixed prefix or only allowing mounting volume roots, we no longer need to have the extra security check.

Reply:
I agree with you!
To solve this problem, I think use key (relative path in the volume) combine with volume mount path is a good solution. 

users can specify volume artifact path as:
```
volume://volume-mount-name/maybe-artifact-or-volume-mount-subPath/artifact/path/here
```

* if `volume-mount-name` not exist or with other `subPath`, throw volume mount not exist error.
* if `volume-mount-name` exist, but no `subPath` specified. return artifact path as:  `mount path` + `/maybe-artifact-or-volume-mount-subPath/artifact/path/here`
* if `volume-mount-name` exist, and `subPath` specified with `maybe-artifact-or-volume-mount-subPath`. return artifact path as:  `mount path`  +  `/artifact/path/here`

If this proposal is ok?

Reply:
@haibingzhao I think our ideas are getting close.
Sorry I didn't quite understand what you meant for `maybe-artifact-or-volume-mount-subPath`.

Can you give some concrete example urls like `volume://volume-mount-name/maybe-artifact-or-volume-mount-subPath/artifact/path/here` with corresponding volumeMount configuration? so that there can be no room for confusion.

Reply:
> @haibingzhao I think our ideas are getting close.
> Sorry I didn't quite understand what you meant for `maybe-artifact-or-volume-mount-subPath`.
> 
> Can you give some concrete example urls like `volume://volume-mount-name/maybe-artifact-or-volume-mount-subPath/artifact/path/here` with corresponding volumeMount configuration? so that there can be no room for confusion.

for example:
1. volume and volume mount config as:
```
spec:
  template:
    spec:
      containers:
        volumeMounts:
          - mountPath: /data1
            subPath: pipeline1
            name: output
          - mountPath: /data2
            subPath: pipeline2
            name: output
          - mountPath: /logs
            name: logs
      volumes:
        - name: output
          persistentVolumeClaim:
                claimName: output,
                readOnly: true
        - name: logs
          persistentVolumeClaim:
                claimName: logs,
                readOnly: true
```
2. if users specify volume artifact path as:
```
volume://output/pipeline1/artifact/path/here
```
and it will match mount name `output`, but there is two mounts named `output`(absolutely they also have subPath field). so check subPath and `pipeline1` matched. finally, real artifact path `/artifact/path/here` returned.

3.  if users specify volume artifact path as:
```
volume://logs/pipeline1/artifact/path/here
```
because only one volume mount `logs`  without subPath matched, real artifact path `/pipeline1/artifact/path/here` returned.

4. other case will throw volume or volume mount not found error.

Reply:
Thanks for the explanation! Yes, I think that's a great idea. Thanks for writing these up.

So for `volume://volume-mount-name/artifact/path/here`, in fact, it doesn't care whether there's subpath or not, it will always record relative path from volume. And for a pod like `ml-pipeline-ui`, it can parse volume mount info to understand real file location based on subPath info.

This is my summary understanding. Is that correct?

Reply:
for `volume://a/b/c/d`, `ml-pipeline-ui` will parse `a`  to volume-mount-name, and will search `ml-pipeline-ui` pod config
    * if there has one volume-mount named `a` and no `subPath` configured, then real path will be combine found volume mountPath (for example `/etc/artifacts`) with `b/c/d`, which is: `/etc/artifacts/b/c/d`
    * if there has one or many volume-mount named `a` with different subPath configured, then `ml-pipeline-ui` will parse `b` to `subPath` to match witch volume-mount-with-subPath to use. if found, real path will be combine volume mountPath (for example `/etc/artifacts`) with `c/d`, which is: `/etc/artifacts/c/d`

In summary, users will care about whether `ml-pipeline-ui` configured `subPath` for each volume mount or not. if `subPath` configured, then `subPath` value will be contained in volume schema path: `volume:/volume-mount-name/sub-path-value/artifact/path/here`. if not, volume schema path will be: `volume:/volume-mount-name/artifact/path/here`

Reply:
sounds good!
Another clarification, `sub-path-value` can be multiple layers, is that right?
It can be `sub/path/value` for example.

Reply:
I think you can start implementation when you have time.

Reply:
> Another clarification, `sub-path-value` can be multiple layers, is that right?
> It can be `sub/path/value` for example.

yes, subPath can be multiple layers.


> I think you can start implementation when you have time.

implementation was done,  waiting for review and suggestion.

Reply:
> implementation was done, waiting for review and suggestion.

Ohh, awesome, I'll take a deeper review tomorrow.